File: app/src/main/java/com/example/core/time/FindOverlapUseCase.kt
package com.example.core.time

import kotlinx.datetime.Instant
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toJavaInstant
import kotlinx.datetime.toKotlinInstant
import java.time.temporal.ChronoUnit
import kotlin.math.min
import kotlin.math.sqrt

/**
 * A participant in a meeting search: a time zone plus that person's working window and an optional
 * do-not-disturb window (§5.8). Defaults model a 9–5 worker with no DND.
 */
data class MeetingParticipant(
    val zoneId: String,
    val workStartHour: Int = 9,
    val workEndHour: Int = 17,
    val dndStartHour: Int = -1,
    val dndEndHour: Int = -1,
)

class FindOverlapUseCase(
    private val timeEngine: TimeEngine
) {

    data class OverlapSlot(
        val utcStartInstant: Instant,
        val localHours: Map<String, Int>,       // ZoneId -> local hour (0-23)
        val localViews: Map<String, LocalView>, // ZoneId -> awake/working/asleep tag
        val totalDiscomfort: Double,
        val standardDeviation: Double,
        val hasDnd: Boolean,
        val ratingLabel: String,                 // "Optimal", "Fair", "Difficult"
        val rankScore: Double                    // Lower is fairer & better
    )

    /** Zone-only convenience: ranks slots treating every zone as a default 9–5 worker. */
    fun calculateBestOverlapSlots(
        baseDateInstant: Instant,
        participantZones: List<TimeZone>
    ): List<OverlapSlot> =
        rankParticipantSlots(baseDateInstant, participantZones.map { MeetingParticipant(it.id) })

    /**
     * Finds and ranks 24 hourly meeting slots over the day starting at [baseDateInstant], scoring
     * each participant against their own working/DND window (§5.8). Lower rankScore = fairer.
     */
    fun rankParticipantSlots(
        baseDateInstant: Instant,
        participants: List<MeetingParticipant>
    ): List<OverlapSlot> {
        if (participants.isEmpty()) return emptyList()

        val startOfWindow = baseDateInstant.toJavaInstant().truncatedTo(ChronoUnit.HOURS)

        val results = mutableListOf<OverlapSlot>()

        for (hourOffset in 0 until 24) {
            val slotInstant = startOfWindow.plusSeconds(hourOffset * 3600L).toKotlinInstant()

            val localHours = mutableMapOf<String, Int>()
            val localViews = mutableMapOf<String, LocalView>()
            val distressList = mutableListOf<Double>()
            var isAnyDnd = false

            for (participant in participants) {
                val zone = TimeZone.of(participant.zoneId)
                val hour = timeEngine.toZonedDateTime(slotInstant, zone).hour
                val distress = discomfortFor(hour, participant)
                distressList.add(distress)
                if (inDndWindow(hour, participant)) isAnyDnd = true

                // Representative view per zone (first participant seen for that zone wins).
                if (participant.zoneId !in localViews) {
                    localHours[participant.zoneId] = hour
                    localViews[participant.zoneId] = viewFor(hour, participant)
                }
            }

            val n = distressList.size
            val sum = distressList.sum()
            val mean = sum / n
            val stdDev = sqrt(distressList.sumOf { (it - mean) * (it - mean) } / n)

            val ratingLabel = when {
                isAnyDnd -> "Difficult"
                sum <= n * 1.5 -> "Optimal"
                else -> "Fair"
            }
            val dndPenalty = if (isAnyDnd) 1000.0 else 0.0
            val rankScore = sum * 10.0 + stdDev + dndPenalty

            results.add(
                OverlapSlot(
                    utcStartInstant = slotInstant,
                    localHours = localHours,
                    localViews = localViews,
                    totalDiscomfort = sum,
                    standardDeviation = stdDev,
                    hasDnd = isAnyDnd,
                    ratingLabel = ratingLabel,
                    rankScore = rankScore
                )
            )
        }

        return results.sortedBy { it.rankScore }
    }

    /** Discomfort cost of a local [hour] for [p]: 0 inside the working window, rising with distance. */
    private fun discomfortFor(hour: Int, p: MeetingParticipant): Double {
        if (inDndWindow(hour, p)) return 8.0
        if (inWindow(hour, p.workStartHour, p.workEndHour)) return 0.0
        return when (hoursOutsideWindow(hour, p.workStartHour, p.workEndHour)) {
            in 0..2 -> 1.5
            in 3..4 -> 3.0
            else -> 6.0
        }
    }

    private fun viewFor(hour: Int, p: MeetingParticipant): LocalView = when {
        inDndWindow(hour, p) -> LocalView.ASLEEP
        inWindow(hour, p.workStartHour, p.workEndHour) -> LocalView.WORKING
        hoursOutsideWindow(hour, p.workStartHour, p.workEndHour) <= 2 -> LocalView.AWAKE
        hoursOutsideWindow(hour, p.workStartHour, p.workEndHour) <= 4 -> LocalView.OUTSIDE_HOURS
        else -> LocalView.ASLEEP
    }

    private fun inDndWindow(hour: Int, p: MeetingParticipant): Boolean =
        p.dndStartHour in 0..23 && p.dndEndHour in 0..23 && inWindow(hour, p.dndStartHour, p.dndEndHour)

    /** True if [hour] falls in [start, end); supports windows that wrap past midnight. */
    private fun inWindow(hour: Int, start: Int, end: Int): Boolean {
        if (start < 0 || end < 0 || start == end) return false
        return if (start < end) hour in start until end else hour >= start || hour < end
    }

    /** Circular distance in hours from [hour] to the nearest edge of the working window. */
    private fun hoursOutsideWindow(hour: Int, start: Int, end: Int): Int {
        if (inWindow(hour, start, end)) return 0
        val lastWorkingHour = (end + 23) % 24 // end is exclusive
        return min(circularDistance(hour, start), circularDistance(hour, lastWorkingHour))
    }

    private fun circularDistance(a: Int, b: Int): Int {
        val diff = ((a - b) % 24 + 24) % 24
        return min(diff, 24 - diff)
    }
}

File: ios/MeridianApp/Core/FindOverlapUseCase.swift
// FindOverlapUseCase.swift
// Meridian — iOS 27 / Swift 6
//
// Finds and ranks fair meeting slots across participants in different time zones.
// Direct behavioral port of Android `FindOverlapUseCase`. Uses MeetingParticipant,
// MeetingSlot, SlotLabel, and LocalView (canonical types in Models.swift / LocalView.swift).
//
// Scoring (must match Android exactly):
//   discomfortFor(hour): 8.0 if in DND; 0.0 if in work window; else by circular distance
//     outside the window: 0..2h -> 1.5, 3..4h -> 3.0, else -> 6.0.
//   totalDiscomfort = sum of per-participant discomfort.
//   standardDeviation = POPULATION std dev of the discomfort list (divide by n).
//   rankScore = totalDiscomfort * 10 + standardDeviation + (anyDnd ? 1000 : 0).
//   label: anyDnd -> .difficult; sum <= n*1.5 -> .optimal; else .fair.
//   Slots are sorted ascending by rankScore (lower is fairer & better).

import Foundation

/// Pure, stateless use-case that finds candidate meeting slots for a set of
/// participants spread across different time zones.
struct FindOverlapUseCase: Sendable {

    init() {}

    // MARK: - Public API

    /// Zone-only convenience: ranks slots treating every zone as a default 9–17 worker.
    /// Mirrors Android `calculateBestOverlapSlots`.
    func calculateBestOverlapSlots(
        baseDate: Date,
        participantZones: [String]
    ) -> [MeetingSlot] {
        rankParticipantSlots(
            baseDate: baseDate,
            participants: participantZones.map { MeetingParticipant(zoneId: $0) }
        )
    }

    /// Finds and ranks 24 hourly meeting slots over the day starting at `baseDate`,
    /// scoring each participant against their own working/DND window.
    /// Lower `rankScore` == fairer. Mirrors Android `rankParticipantSlots`.
    func rankParticipantSlots(
        baseDate: Date,
        participants: [MeetingParticipant]
    ) -> [MeetingSlot] {
        guard !participants.isEmpty else { return [] }

        // Android: baseDateInstant.truncatedTo(ChronoUnit.HOURS) — floor to the top of the hour.
        let startOfWindow = floorToHour(baseDate)

        var results: [MeetingSlot] = []
        results.reserveCapacity(24)

        for hourOffset in 0..<24 {
            let slotInstant = startOfWindow.addingTimeInterval(Double(hourOffset) * 3600.0)

            var localHours: [String: Int] = [:]
            var localViews: [String: LocalView] = [:]
            var distressList: [Double] = []
            distressList.reserveCapacity(participants.count)
            var isAnyDnd = false

            for participant in participants {
                let hour = localHour(at: slotInstant, in: participant.zoneId)
                let distress = discomfortFor(hour: hour, participant)
                distressList.append(distress)
                if inDndWindow(hour: hour, participant) { isAnyDnd = true }

                // Representative view per zone (first participant seen for that zone wins).
                if localViews[participant.zoneId] == nil {
                    localHours[participant.zoneId] = hour
                    localViews[participant.zoneId] = viewFor(hour: hour, participant)
                }
            }

            let n = Double(distressList.count)
            let sum = distressList.reduce(0.0, +)
            let mean = sum / n
            let variance = distressList.reduce(0.0) { acc, value in
                let delta = value - mean
                return acc + delta * delta
            } / n
            let stdDev = variance.squareRoot()

            let label: SlotLabel
            if isAnyDnd {
                label = .difficult
            } else if sum <= n * 1.5 {
                label = .optimal
            } else {
                label = .fair
            }

            let dndPenalty = isAnyDnd ? 1000.0 : 0.0
            let rankScore = sum * 10.0 + stdDev + dndPenalty

            results.append(
                MeetingSlot(
                    start: slotInstant,
                    localHours: localHours,
                    localViews: localViews,
                    totalDiscomfort: sum,
                    standardDeviation: stdDev,
                    hasDnd: isAnyDnd,
                    rankScore: rankScore,
                    label: label
                )
            )
        }

        return results.sorted { $0.rankScore < $1.rankScore }
    }

    // MARK: - Scoring (private, mirrors Android)

    /// Discomfort cost of a local `hour`: 0 inside the working window, rising with distance.
    private func discomfortFor(hour: Int, _ p: MeetingParticipant) -> Double {
        if inDndWindow(hour: hour, p) { return 8.0 }
        if inWindow(hour: hour, start: p.workStartHour, end: p.workEndHour) { return 0.0 }
        switch hoursOutsideWindow(hour: hour, start: p.workStartHour, end: p.workEndHour) {
        case 0...2: return 1.5
        case 3...4: return 3.0
        default:    return 6.0
        }
    }

    private func viewFor(hour: Int, _ p: MeetingParticipant) -> LocalView {
        if inDndWindow(hour: hour, p) { return .asleep }
        if inWindow(hour: hour, start: p.workStartHour, end: p.workEndHour) { return .working }
        let outside = hoursOutsideWindow(hour: hour, start: p.workStartHour, end: p.workEndHour)
        if outside <= 2 { return .awake }
        if outside <= 4 { return .outsideHours }
        return .asleep
    }

    private func inDndWindow(hour: Int, _ p: MeetingParticipant) -> Bool {
        (0...23).contains(p.dndStartHour) &&
            (0...23).contains(p.dndEndHour) &&
            inWindow(hour: hour, start: p.dndStartHour, end: p.dndEndHour)
    }

    /// True if `hour` falls in `[start, end)`; supports windows that wrap past midnight.
    private func inWindow(hour: Int, start: Int, end: Int) -> Bool {
        if start < 0 || end < 0 || start == end { return false }
        if start < end {
            return hour >= start && hour < end
        } else {
            return hour >= start || hour < end
        }
    }

    /// Circular distance in hours from `hour` to the nearest edge of the working window.
    private func hoursOutsideWindow(hour: Int, start: Int, end: Int) -> Int {
        if inWindow(hour: hour, start: start, end: end) { return 0 }
        let lastWorkingHour = (end + 23) % 24 // end is exclusive
        return min(circularDistance(hour, start), circularDistance(hour, lastWorkingHour))
    }

    private func circularDistance(_ a: Int, _ b: Int) -> Int {
        let diff = ((a - b) % 24 + 24) % 24
        return min(diff, 24 - diff)
    }

    // MARK: - Time helpers

    /// Floors a `Date` to the top of its UTC hour (Android: truncatedTo(HOURS)).
    private func floorToHour(_ date: Date) -> Date {
        let seconds = date.timeIntervalSinceReferenceDate
        let floored = (seconds / 3600.0).rounded(.down) * 3600.0
        return Date(timeIntervalSinceReferenceDate: floored)
    }

    /// Local hour (0–23) at a given absolute instant for an IANA time-zone identifier.
    /// Falls back to UTC when the identifier is unrecognized.
    private func localHour(at date: Date, in tzId: String) -> Int {
        let tz = TimeZone(identifier: tzId) ?? TimeZone(identifier: "UTC")!
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = tz
        return cal.component(.hour, from: date)
    }
}
